socketioxide/src/retryer.rs

```diff
@@ -0,0 +1,139 @@
+use crate::errors::RetryerError;
+use engineioxide::{sid_generator::Sid, SendPacket};
+use std::{collections::VecDeque, fmt::Debug};
+use tokio::{sync::mpsc::error::TrySendError, sync::mpsc::Sender};
+
+// todo bin payload and payload should be in one VecDeque,
+// todo sender should be the same for both types of SendPacket
+
+/// The `Retryer` struct represents a retry mechanism for sending packets.
+#[derive(Debug)]
+pub struct Retryer<T: Debug> {
+    sid: Sid,
+    sender: Sender<T>,
+    packet: Option<T>,
+    bin_payload: VecDeque<Vec<u8>>,
+    bin_sender: Sender<SendPacket>,
+}
+
+impl<T: Debug> Retryer<T> {
+    /// Creates a new `Retryer` instance with the specified parameters.
+    ///
+    /// # Arguments
+    ///
+    /// * `sid` - The identifier for the network connection.
+    /// * `sender` - The sender channel used for sending packets.
+    /// * `packet` - An optional packet to be sent initially.
+    /// * `bin_payload` - A queue of binary payloads to be sent.
+    /// * `bin_sender` - The sender channel used for sending binary packets.
+    ///
+    /// # Returns
+    ///
+    /// A new `Retryer` instance.
+    pub(crate) fn new(
+        sid: Sid,
+        sender: Sender<T>,
+        packet: Option<T>,
+        bin_payload: VecDeque<Vec<u8>>,
+        bin_sender: Sender<SendPacket>,
+    ) -> Self {
+        Self {
+            sid,
+            sender,
+            packet,
+            bin_payload,
+            bin_sender,
+        }
+    }
+
+    /// Retries sending the packet and binary payloads.
+    ///
+    /// # Returns
+    ///
+    /// - `Ok(())` if all packets were successfully sent.
+    /// - `Err(RetryerError)` if there are remaining packets to be sent or the socket is closed.
+    pub fn retry(mut self) -> Result<(), RetryerError<T>> {
+        // Retry sending the main packet
+        match self.packet.map(|p| self.sender.try_send(p)) {
```

Comment:
Here you can do self.packet.take() to take the ownership of the packet.
You can then reuse the retryer in case of error whithout recreating one : 
```rust
self.packet = Some(packet);
return Err(RetryerError::Remaining(self))
```